consensus/core/src/tests/randomized_tests.rs

```diff
@@ -0,0 +1,233 @@
+        let leader_round = i as u32 + 1;
+        if let DecidedLeader::Commit(ref block) = leader_block {
+            assert_eq!(block.round(), leader_round);
+            assert_eq!(
+                block.author(),
+                authority.committer.get_leaders(leader_round)[0]
+            );
+        } else {
+            panic!("Expected a committed leader")
+        };
+    }
+}
+
+/// Test builds a randomized dag with the following conditions:
+/// - Links to 2f+1 minimum ancestors
+/// - Links to leader of previous round 50% of the time.
+///
+/// Blocks will randomly be fed through BlockManager and after each accepted
+/// block we will try_decide() and if there is a committed sequence we will update
+/// last_decided and continue. We do this from the perspective of two different
+/// authorities who receive the blocks in different orders and ensure the resulting
+/// sequence is the same for both authorities. The resulting sequence will include
+/// Commit & Skip decisions and potentially will stop before coming to a decision
+/// on all waves as we may have an Undecided leader somewhere early in the sequence.
+#[test]
+fn test_randomized_dag_and_decision_sequence() {
+    let mut random_test_setup = random_test_setup();
+
+    // Setup for Authority 1
+    let mut authority_1 = authority_setup(random_test_setup.num_authorities);
+
+    let pipeline = random_test_setup.num_waves % DEFAULT_WAVE_LENGTH as usize;
+    let wave_number =
+        authority_1.committer.committers[pipeline].wave_number(random_test_setup.num_waves as u32);
+    let num_rounds = authority_1.committer.committers[pipeline].decision_round(wave_number);
+    let include_leader_percentage = 50;
+    let dag_builder = create_random_dag(
+        random_test_setup.seed,
+        include_leader_percentage,
+        num_rounds,
+        authority_1.context.clone(),
+    );
+
+    tracing::info!(
+        "Running test with committee size {} & {} completed waves in the DAG...",
+        random_test_setup.num_authorities,
+        random_test_setup.num_waves
+    );
+
+    let mut all_blocks = dag_builder.blocks.values().cloned().collect::<Vec<_>>();
+    all_blocks.shuffle(&mut random_test_setup.seeded_rng);
+
+    let mut sequenced_leaders_1 = vec![];
+    let mut last_decided = Slot::new_for_test(0, 0);
+    for block in &all_blocks {
+        let _ = authority_1
+            .block_manager
+            .try_accept_blocks(vec![block.clone()]);
+        let sequence = authority_1.committer.try_decide(last_decided);
+
+        if !sequence.is_empty() {
+            sequenced_leaders_1.extend(sequence.clone());
+            let leader_status = sequence.last().unwrap();
+            last_decided = Slot::new(leader_status.round(), leader_status.authority());
+        }
+    }
+
+    assert!(authority_1.block_manager.is_empty());
+
+    // Setup for Authority 2
+    let mut authority_2 = authority_setup(random_test_setup.num_authorities);
```

Comment:
Nice. Is it possible to use an env var to control running the rest of logic multiple times?